### PR TITLE
Data: New - Custom error handling.

### PR DESCRIPTION
**Pull request by iNewLegend**

### CODE REVIEW

File: data/error-manager.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
+<?php
+namespace Elementor\Data;
+
```
Comment (iNewLegend) [inline data/error-manager.php:3]:
```suggestion
// TODO: Remove this file, and create global error manager.

```

File: data/manager.php [hunk 2]
```diff
@@ -5,27 +5,21 @@
 use Elementor\Core\Base\Module as BaseModule;
 use Elementor\Data\Base\Controller;
 use Elementor\Data\Base\Processor;
-use Elementor\Data\Editor;
 
 if ( ! defined( 'ABSPATH' ) ) {
```
Comment (iNewLegend) [inline data/manager.php:9]:
```suggestion

// TODO: Remove this file, and create global error manager.

if ( ! defined( 'ABSPATH' ) ) {
```

File: data/error-manager.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
+<?php
```
Comment (matipojo) [inline data/error-manager.php:1]: See `core/logger/manager.php`,  it should be implemented there as a logger: `core/logger/loggers/base.php`.

File: data/manager.php [hunk 4]
```diff
@@ -41,6 +35,30 @@ public static function instance() {
...
+		add_action( 'rest_api_init', function () {
+			// Empty server means: this is not an internal run.
+			if ( ! $this->is_internal && ! Error_Manager::is_registered() ) {
+				Error_Manager::register_handler( [ $this, 'rest_error_handler' ] );
+			}
+		} );
+	}
+
+	public function __destruct() {
+		if ( Error_Manager::is_registered() ) {
```
Comment (matipojo) [inline data/manager.php:57]: Check `is_registered` from within the manager.

File: data/manager.php [hunk 5]
```diff
@@ -41,6 +35,30 @@ public static function instance() {
...
+	 */
+	public static function instance() {
+		return ( parent::instance() );
+	}
+
+	public function __construct() {
+		add_action( 'rest_api_init', function () {
+			// Empty server means: this is not an internal run.
+			if ( ! $this->is_internal && ! Error_Manager::is_registered() ) {
+				Error_Manager::register_handler( [ $this, 'rest_error_handler' ] );
```
Comment (matipojo) [inline data/manager.php:51]: Check `is_registered` from within the manager.

File: data/manager.php [hunk 6]
```diff
@@ -194,6 +212,8 @@ public function command_to_endpoint( $command, $format, $args ) {
 	 * @return \WP_REST_Server
 	 */
 	public function run_server() {
+		$this->is_internal = true;
```
Comment (matipojo) [inline data/manager.php:215]: Explain.

File: data/manager.php [hunk 7]
```diff
@@ -350,4 +371,19 @@ public function run( $command, $args = [], $method = 'GET' ) {
 
 		return $result;
 	}
+
+	/**
+	 * @param \WP_Error $error
+	 */
+	public function rest_error_handler( $error ) {
```
Comment (matipojo) [inline data/manager.php:378]: Move to logger.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: data/manager.php [hunk 8]
```diff
@@ -41,6 +35,30 @@ public static function instance() {
...
+	 *
+	 * @return \Elementor\Core\Base\Module|\Elementor\Data\Manager
+	 */
+	public static function instance() {
+		return ( parent::instance() );
+	}
+
+	public function __construct() {
+		add_action( 'rest_api_init', function () {
+			// Empty server means: this is not an internal run.
```
Comment (iNewLegend) [inline data/manager.php:49]: fix comment

File: data/error-manager.php (hunk 3 above)
Comment (iNewLegend) [reply data/error-manager.php]: Not relevant anymore.

File: data/manager.php (hunk 4 above)
Comment (iNewLegend) [reply data/manager.php]: Not relevant anymore.

File: data/manager.php (hunk 5 above)
Comment (iNewLegend) [reply data/manager.php]: Not relevant anymore.

File: data/manager.php (hunk 6 above)
Comment (iNewLegend) [reply data/manager.php:203]: Handled.

File: data/manager.php (hunk 7 above)
Comment (iNewLegend) [reply data/manager.php]: Handled.

File: core/logger/manager.php [hunk 9]
```diff
@@ -50,6 +52,40 @@ public function shutdown() {
...
+			}
+		}
+
+		$this->shutdown( $error->get_error_data() );
+
+		exit;
+	}
+
+	public function register_rest_error_handler() {
+		if ( ! \Elementor\Data\Manager::instance()->is_internal() ) {
```
Comment (matipojo) [inline core/logger/manager.php:84]: The data manger should register and use this manger.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/logger/manager.php (hunk 9 above)
Comment (iNewLegend) [reply core/logger/manager.php]: Handled: https://github.com/elementor/elementor/pull/11853/commits/14aee2dd7f9f201c63f393c14f54395f47e04a21

File: core/logger/manager.php [hunk 10]
```diff
@@ -50,6 +52,34 @@ public function shutdown() {
...
+			if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
+				echo wp_json_encode( $error->get_error_data() );
+			} else {
+				echo wp_json_encode( [
+					'message' => 'Server error, see Elementor => System Info',
+				] );
+			}
+		}
+
+		$this->shutdown( $error->get_error_data() );
```
Comment (matipojo) [inline core/logger/manager.php:78]: It doesn't called automatically?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/logger/manager.php (hunk 10 above)
Comment (iNewLegend) [reply core/logger/manager.php:78]: does not work via shutdown.

**Review by matipojo — APPROVED** [review]

**Review by KingYes — APPROVED** [review]

**Pull request merged**